File: shared/java/com/jeffgeissler/retroeightball/Responses.java
package com.jeffgeissler.retroeightball;

import java.util.Random;

/** Original answer order and weights, shared by desktop and Android. */
public final class Responses {
    private static final String[] ANSWERS = {
        "It is certain.", "Ask again later.", "My reply is no.", "Outlook good.",
        "Better not tell you now.", "Yes — definitely.", "Very doubtful.", "Signs point to yes."
    };
    private static final int[] WEIGHTS = {3, 2, 3, 4, 1, 5, 2, 4};
    public static final int TOTAL_WEIGHT = 24;
    private Responses() {}

    public static String choose(Random random) { return forRoll(random.nextInt(TOTAL_WEIGHT)); }

    /** Maps a zero-based roll to an answer; useful for deterministic verification. */
    public static String forRoll(int roll) {
        if (roll < 0 || roll >= TOTAL_WEIGHT) throw new IllegalArgumentException("Roll must be 0–23");
        for (int i = 0; i < WEIGHTS.length; i++) {
            if (roll < WEIGHTS[i]) return ANSWERS[i];
            roll -= WEIGHTS[i];
        }
        throw new AssertionError("Invalid response weights");
    }
}

File: ios/RetroEightBall/Responses.swift
enum Responses {
    static let answers: [(text: String, weight: Int)] = [
        ("It is certain.", 3), ("Ask again later.", 2), ("My reply is no.", 3),
        ("Outlook good.", 4), ("Better not tell you now.", 1), ("Yes — definitely.", 5),
        ("Very doubtful.", 2), ("Signs point to yes.", 4)
    ]
    static let totalWeight = answers.reduce(0) { $0 + $1.weight }
    static func answer(for roll: Int) -> String {
        precondition((0..<totalWeight).contains(roll))
        var remainder = roll
        for answer in answers {
            if remainder < answer.weight { return answer.text }
            remainder -= answer.weight
        }
        preconditionFailure("Invalid response weights")
    }
    static func choose() -> String { answer(for: Int.random(in: 0..<totalWeight)) }
}
